### PR TITLE
DD-746 Slug Improved

### PR DESCRIPTION
**Pull request by marcelovani**

### CODE REVIEW

File: src/Form/AdminSettings.php [hunk 1]
```diff
@@ -155,13 +155,24 @@ public function buildForm(array $form, FormStateInterface $form_state, Request $
...
+      '#options' => [
+        $this->t('Above the ad container'),
+        $this->t('Inside the ad container'),
+      ],
+      '#description' => $this->t('Please note that in case <em>refresh()</em> method is used, slug placed inside the slot container will disappear.'),
+    ];
     $form['global_display_options']['hide_slug'] = [
       '#type' => 'checkbox',
-      '#title' => $this->t('Hide slug if no ad is served (recommended)'),
+      '#title' => $this->t('Only show slug when the Ad is served (recommended)'),
```
Comment (marcelovani) [inline src/Form/AdminSettings.php:170]: Improved message

File: dfp.module [hunk 2]
```diff
@@ -47,6 +47,8 @@ function dfp_theme() {
         'single_request' => NULL,
         'collapse_empty_divs' => NULL,
         'disable_init_load' => NULL,
+        'slug_position' => NULL,
```
Comment (marcelovani) [inline dfp.module:50]: Passing these variables to the template will help to decide when to add the scripts

File: src/Tests/DisplayTagTest.php [hunk 3]
```diff
@@ -103,16 +103,27 @@ public function testSlug() {
     $edit['slug'] = 'Tag specific slug';
     $this->dfpEditTag($tag->id(), $edit);
     $this->drupalGet('<front>');
-    $this->assertText('Tag specific slug');
-
-    // Set the slug to be hidden. Use admin UI and the cache tags added in
+    $link = $this->xpath('//div[contains(@class, :c)]//div[contains(@class, :sc) and not(contains(@class, "hidden"))][contains(string(), :st)]/following-sibling::script', [
```
Comment (marcelovani) [inline src/Tests/DisplayTagTest.php:106]: Tests for the visibility and position

File: src/Tests/DisplayTagTest.php [hunk 4]
```diff
@@ -103,16 +103,27 @@ public function testSlug() {
...
     $this->drupalPostForm('admin/structure/dfp/settings', $edit, t('Save configuration'));
     $this->drupalGet('<front>');
-    $this->assertNoText('Tag specific slug');
+    $link = $this->xpath('//div[contains(@class, :c)]//div[contains(@class, :sc) and contains(@class, "hidden")][contains(string(), :st)]/following-sibling::div', [
+      ':c' => 'block-dfp',
+      ':sc' => 'dfp-tag--slug',
+      ':st' => 'Tag specific slug'
+    ]);
+    $this->assertTrue(count($link) === 1, 'The slug is hidden above the Ad container');
   }
```
Comment (marcelovani) [inline src/Tests/DisplayTagTest.php]: Tests for the visibility and position

File: src/Tests/DisplayTagTest.php [hunk 5]
```diff
@@ -103,16 +103,27 @@ public function testSlug() {
...
+    // Set slug the above the Ad and hidden. Use admin UI and the cache tags added in
     // \Drupal\dfp\View\TagViewBuilder::viewMultiple() are tested.
     $edit = [
       'hide_slug' => TRUE,
+      'slug_position' => 0,
     ];
     $this->drupalPostForm('admin/structure/dfp/settings', $edit, t('Save configuration'));
     $this->drupalGet('<front>');
-    $this->assertNoText('Tag specific slug');
+    $link = $this->xpath('//div[contains(@class, :c)]//div[contains(@class, :sc) and contains(@class, "hidden")][contains(string(), :st)]/following-sibling::div', [
```
Comment (marcelovani) [inline src/Tests/DisplayTagTest.php:121]: Tests for the visibility and position

File: src/View/TagView.php [hunk 6]
```diff
@@ -401,4 +401,15 @@ public function isSlugHidden() {
...
   }
 
+  /**
+   * Gets the slug position.
+   *
+   * @return integer
+   *   0: Above
+   *   1: Inside
+   */
+  public function getSlugPosition() {
```
Comment (marcelovani) [inline src/View/TagView.php:411]: generic function

File: templates/dfp-js-head-bottom.html.twig [hunk 7]
```diff
@@ -41,5 +43,16 @@ googletag.cmd.push(function() {
     ]);
   {% endfor %}
 });
+{% if slug_position == 0 and hide_slug %}
```
Comment (marcelovani) [inline templates/dfp-js-head-bottom.html.twig:46]: Check position and hide status

File: templates/dfp-tag.html.twig [hunk 8]
```diff
@@ -11,15 +11,23 @@
  * @ingroup themeable
  */
 #}
+{% set slugElement %}
```
Comment (marcelovani) [inline templates/dfp-tag.html.twig:14]: Prepare element

**Comment by simon-peacock** [comment]
Experimental - closing

**Pull request closed without merging**